cdap-gateway/src/main/java/io/cdap/cdap/gateway/router/handlers/HttpRequestRouter.java

```diff
@@ -192,6 +192,30 @@ public void channelInactive(ChannelHandlerContext ctx) {
     ctx.fireChannelInactive();
   }
 
+  /**
+   * [CDAP-21071] Handles the case by stopping the outbound channel of Service -> Router
+   * when the response from Service -> Router is faster than the response from Router to the Client.
+   */
+  @Override
+  public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {
+    if (inflightRequests > 0 && currentMessageSender.outboundChannel != null) {
+      final Channel outboundChannel = ctx.channel();
+      ctx.executor().execute(() -> {
+        // If outboundChannel is not saturated anymore, continue accepting
+        // the incoming traffic from the outbound channel for service<>router.
+        if (outboundChannel.isWritable()) {
+          LOG.trace("Setting message sender's outboundChannel readable.");
```

Comment:
Create a method in MessageSender. Mutating `currentMessageSender.outboundChannel` state directly leaks abstraction

Reply:
nit: move trace logs over there as well. This would allow you to simply do `currentMessageSender.setAutoRead(inboundChannel.isWritable())`